JOSESwift/Sources/Decrypter.swift

```diff
@@ -70,22 +71,45 @@ public struct Decrypter {
     ///
     /// - Parameters:
     ///   - keyDecryptionAlgorithm: The algorithm used to decrypt the shared content encryption key.
-    ///   - kdk: The private key used to decrypt the shared content encryption key.
-    ///          Currently supported key types are: `SecKey`.
+    ///   - key: The key used to perform the decryption. If the `keyDecryptionAlgorithm` is `.direct`, the
+    ///          `decryptionKey` is the shared symmetric content encryption key. Otherwise the `decryptionKey` is the
+    ///           private key of the receiver. See [RFC-7516](https://tools.ietf.org/html/rfc7516#section-5.2) for
+    ///           details.
     ///   - contentDecryptionAlgorithm: The algorithm used to decrypt the JWE's payload.
     /// - Returns: A fully initialized `Decrypter` or `nil` if provided key is of the wrong type.
-    public init?<KeyType>(keyDecryptionAlgorithm: AsymmetricKeyAlgorithm, keyDecryptionKey kdk: KeyType, contentDecryptionAlgorithm: SymmetricKeyAlgorithm) {
+    public init?<KeyType>(keyDecryptionAlgorithm: AsymmetricKeyAlgorithm, decryptionKey key: KeyType, contentDecryptionAlgorithm: SymmetricKeyAlgorithm) {
         switch (keyDecryptionAlgorithm, contentDecryptionAlgorithm) {
         case (.RSA1_5, .A256CBCHS512):
-            guard type(of: kdk) is RSADecrypter.KeyType.Type else {
+            guard type(of: key) is RSADecrypter.KeyType.Type else {
```

Comment:
That's a very good point. I did some testing and yes, normally that would work:

``` swift
func printType<T>(of key: T) {
    if let _ = key as? Data {
        print("Data")
    }
}

printType(of: Data()) // prints "Data"
``` 

However, doing the same with `SecKey` (`if let _ = key as? SecKey`) gives the following error:

> error: conditional downcast to CoreFoundation type 'SecKey' will always succeed

After more testing, this seems to be the situation:

- For normal Swift types we could do optional let bindings as you suggested
- The current check `type(of: key) is SecKey` will succeed for all CoreFoundation types (SecKey, CFString, ...) but not for Swift types.

So if one passes `Data()` as RSA key it will detect the wrong type, but if one passes e.g. a CFString, it won't detect the error. The correct way of checking CoreFoundation types would be to compare `CFGetTypeID(key as CFTypeRef` to stuff like `SecKeyGetTypeID()`. However I currently don't know how we should solve that without knowing what `*GetTypeID()` function to call (since the `KeyType` is generic and `CFGetTypeID()` only seems to work on `AnyObject`s  and not on types iteself).

Since that's a bug not directly introduced by this pull request and it does not affect direct encryption, are you guys ok with fixing it separately? If yes, I'll track an issue. cc @carol-mohemian, @gigi-mohemian

Reply:
Thanks a lot for the thorough explanation! 

Fixing this separately as you suggested is the way to go I'd say 👍